pytorch_lightning/trainer/trainer.py

```diff
@@ -1141,7 +1141,7 @@ def call_teardown_hook(self, model: LightningModule) -> None:
         state = self._teardown_state
 
         if self.datamodule is not None:
-            called = getattr(self.datamodule, f'has_teardown_{state}')
+            called = getattr(self.datamodule, f'has_teardown_{state}', False)
             if not called:
```

Comment:
Should we move this `getattr` check to avoid the function call inside here? (also for `setup`):

https://github.com/PyTorchLightning/pytorch-lightning/blob/f48ac62334e557f694d411e704a5da5d5e58d69a/pytorch_lightning/core/datamodule.py#L55-L97

Because we have the bug here where if stage is `None` we are not checking if has already been called.

We could also actually add the `has_teardown_None` attribute.